CYPRESS_ env variable string value containing commas parsed as multiple values

<!-- Is this a question? Questions WILL BE CLOSED. Ask in our chat https://on.cypress.io/chat  -->

### Current behavior
After upgrading Cypress from 4.10.0 to 5.3.0, one of my tests started failing. This appears to be due to a change in the way `CYPRESS_` environment variables are getting parsed between the two versions. 

I'm passing a string of JSON data to my test via a `CYPRESS_` prefixed environment variable like so:

```
CYPRESS_stringified_json='[{"type": "foo", "value": "bar"}, {"type": "fizz", "value": "buzz"}]' node_modules/.bin/cypress run --spec cypress/integration/spec.js
```

When attempting to call `JSON.parse(Cypress.env('stringified_json'))` on that value in my test, the test started failing with the following error:

```
 > Unexpected token , in JSON at position ###
```

After digging in a little bit, I found that instead of the expected JSON string, the value for that variable now contains an array of the passed string value split on every comma. Here's the output for 5.3.0 using the simplified test code attached below:

```
====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    5.3.0                                                                              │
  │ Browser:    Electron 83 (headless)                                                             │
  │ Specs:      1 found (spec.js)                                                                  │
  │ Searched:   cypress/integration/spec.js                                                        │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  spec.js                                                                         (1 of 1)


  process environment variables
    1) correctly parses stringified JSON data


  0 passing (309ms)
  1 failing

  1) process environment variables
       correctly parses stringified JSON data:

      AssertionError: expected [ Array(4) ] to equal '[{"type": "foo", "value": "bar"}, {"type": "fizz", "value": "buzz"}]'
      + expected - actual

      -[ '{"type": "foo"',
      -  ' "value": "bar"}',
      -  ' {"type": "fizz"',
      -  ' "value": "buzz"}' ]
      +'[{"type": "foo", "value": "bar"}, {"type": "fizz", "value": "buzz"}]'
      
      at Context.eval (http://localhost:34691/__cypress/tests?p=cypress/integration/spec.js:101:48)




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        1                                                                                │
  │ Passing:      0                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  1                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     0 seconds                                                                        │
  │ Spec Ran:     spec.js                                                                          │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /cypress/screenshots/spec.js/process environment variables     (1280x720)
      -- correctly parses stringified JSON data (failed).png                                        


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✖  spec.js                                  300ms        1        -        1        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  1 of 1 failed (100%)                     300ms        1        -        1        -        -  

```

### Desired behavior
Simply pass through the value for the environment variable without splitting on commas the same way it used to be passed in 4.10.0. Here's the output of the test code attached below for 4.10.0:

```
====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    4.10.0                                                                             │
  │ Browser:    Electron 80 (headless)                                                             │
  │ Specs:      1 found (spec.js)                                                                  │
  │ Searched:   cypress/integration/spec.js                                                        │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────

  Running:  spec.js                                                                         (1 of 1)


  0 passing (1ms)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        0                                                                                │
  │ Passing:      0                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     0 seconds                                                                        │
  │ Spec Ran:     spec.js                                                                          │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  spec.js                                    1ms        -        -        -        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
	✔  All specs passed!                          1ms        -        -        -        -        -
```

### Test code to reproduce
```js
// CYPRESS_stringified_json='[{"type": "foo", "value": "bar"}, {"type": "fizz", "value": "buzz"}]' node_modules/.bin/cypress run --spec cypress/integration/spec.js

describe('process environment variables', () => {
	it('correctly parses stringified JSON data', () => {
		expect(Cypress.env('stringified_json')).to.equal(
			'[{"type": "foo", "value": "bar"}, {"type": "fizz", "value": "buzz"}]'
		);
	});
});
```

### Versions
Cypress: Cypress 5.3.0
Operating System: macOS 10.15.7
Browser: Electron 83 (headless)

--- a/packages/server/lib/util/args.js
+++ b/packages/server/lib/util/args.js
@@ -243,13 +243,19 @@ module.exports = {
         return path.resolve(options.cwd, p)
       }
 
-      // clean up single quotes wrapping the spec for Windows users
-      // https://github.com/cypress-io/cypress/issues/2298
-      if (spec[0] === '\'' && spec[spec.length - 1] === '\'') {
-        spec = spec.substring(1, spec.length - 1)
-      }
+      // https://github.com/cypress-io/cypress/issues/8818
+      // Sometimes spec is parsed to array. Because of that, we need check.
+      if (typeof spec === 'string') {
+        // clean up single quotes wrapping the spec for Windows users
+        // https://github.com/cypress-io/cypress/issues/2298
+        if (spec[0] === '\'' && spec[spec.length - 1] === '\'') {
+          spec = spec.substring(1, spec.length - 1)
+        }
 
-      options.spec = strToArray(spec).map(resolvePath)
+        options.spec = strToArray(spec).map(resolvePath)
+      } else {
+        options.spec = spec.map(resolvePath)
+      }
     }
 
     if (tag) {

--- a/packages/server/lib/util/args.js
+++ b/packages/server/lib/util/args.js
@@ -125,8 +125,12 @@ const sanitizeAndConvertNestedArgs = (str, argname) => {
   la(is.unemptyString(argname), 'missing config argname to be parsed')
 
   try {
-  // if this is valid JSON then just
-  // parse it and call it a day
+    if (typeof str === 'object') {
+      return str
+    }
+
+    // if this is valid JSON then just
+    // parse it and call it a day
     const parsed = tryJSONParse(str)
 
     if (parsed) {

--- a/packages/server/lib/util/args.js
+++ b/packages/server/lib/util/args.js
@@ -259,7 +259,7 @@ module.exports = {
     }
 
     if (tag) {
-      options.tag = strToArray(tag)
+      options.tag = typeof tag === 'string' ? strToArray(tag) : tag
     }
 
     if (env) {

--- a/packages/server/lib/util/coerce.js
+++ b/packages/server/lib/util/coerce.js
@@ -51,7 +51,7 @@ module.exports = (value) => {
 
   const obj = fromJson(value)
 
-  if (obj) {
+  if (obj && typeof obj === 'object') {
     return obj
   }
 

--- a/packages/server/lib/util/coerce.js
+++ b/packages/server/lib/util/coerce.js
@@ -1,12 +1,6 @@
 const _ = require('lodash')
 const toBoolean = require('underscore.string/toBoolean')
 
-const isValue = (value) => {
-  return (comparison) => {
-    return _.invoke(comparison, 'toString') === value
-  }
-}
-
 // https://github.com/cypress-io/cypress/issues/6810
 const toArray = (value) => {
   const valueIsNotStringOrArray = typeof (value) !== 'string' || (value[0] !== '[' && value[value.length - 1] !== ']')
@@ -30,14 +24,42 @@ const toArray = (value) => {
   return arr
 }
 
+// https://github.com/cypress-io/cypress/issues/8818
+// toArray() above doesn't handle JSON string properly.
+// For example, '[{a:b,c:d},{e:f,g:h}]' isn't the parsed object but ['{a:b', 'c:d}', '{e:f', 'g:h}']. It's useless.
+// Because of that, we check if the value is a JSON string.
+const fromJson = (value) => {
+  try {
+    return JSON.parse(value)
+  } catch (e) {
+    // do nothing
+  }
+}
+
 module.exports = (value) => {
   const num = _.toNumber(value)
+
+  if (_.invoke(num, 'toString') === value) {
+    return num
+  }
+
   const bool = toBoolean(value)
+
+  if (_.invoke(bool, 'toString') === value) {
+    return bool
+  }
+
+  const obj = fromJson(value)
+
+  if (obj) {
+    return obj
+  }
+
   const arr = toArray(value)
 
-  return _
-  .chain([num, bool, arr])
-  .find(isValue(value))
-  .defaultTo(value)
-  .value()
+  if (_.invoke(arr, 'toString') === value) {
+    return arr
+  }
+
+  return value
 }

--- a/packages/server/test/unit/coerce_spec.js
+++ b/packages/server/test/unit/coerce_spec.js
@@ -48,6 +48,23 @@ describe('lib/util/coerce', () => {
       expect(coerce(cypressEnvVar)).to.deep.include({ BOOLEAN: false })
     })
 
+    // https://github.com/cypress-io/cypress/issues/8818
+    it('coerces JSON string', () => {
+      expect(coerce('[{"type": "foo", "value": "bar"}, {"type": "fizz", "value": "buzz"}]')).to.deep.equal(
+        [{ 'type': 'foo', 'value': 'bar' }, { 'type': 'fizz', 'value': 'buzz' }],
+      )
+    })
+
+    // https://github.com/cypress-io/cypress/issues/8818
+    it('coerces JSON string from process.env', () => {
+      process.env['CYPRESS_stringified_json'] = '[{"type": "foo", "value": "bar"}, {"type": "fizz", "value": "buzz"}]'
+      const cypressEnvVar = getProcessEnvVars(process.env)
+      const coercedCypressEnvVar = coerce(cypressEnvVar)
+
+      expect(coercedCypressEnvVar).to.have.keys('stringified_json')
+      expect(coercedCypressEnvVar['stringified_json']).to.deep.equal([{ 'type': 'foo', 'value': 'bar' }, { 'type': 'fizz', 'value': 'buzz' }])
+    })
+
     it('coerces array', () => {
       expect(coerce('[foo,bar]')).to.have.members(['foo', 'bar'])
     })

